scripts/train_lm.py

```diff
@@ -144,15 +144,105 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
+    # this case ca be very rare)
+    for i in range(0, size): 
+        if out[i] in [0.0, 1.0]:
+            out[i] = RoundToProperDigit(metaparameters[i], round_index)
+            if out[i] in [0.0, 1.0]:
+                sys.exit("train_lm.py: {0}th parameter is {1}. metaparameters should be in range (0, 1)".format(i + 1, out[i]))
+    
+    # append the first #num_train_sets parameters to the output set 
+    # note: the first #num_train_sets parameters (not data-type weights) do not need to be checked
+    assert num_train_sets >= 1 and num_train_sets < len(metaparameters)
+    for param in out[:num_train_sets]:
+        out_param.append(format(param))
+   
+    # check repeating value in the rest parameters. 
+    # after this while loop, there should be no repeating values in set out
+    (index1, index2) = FindRepeatingValue(out[num_train_sets:])
+    while (index1 != -1):
+        i = index1 + num_train_sets
+        j = index2 + num_train_sets
+        (updated_param1, updated_param2) = AdjustValues(out[i], out[j], i, j, metaparameters[i], metaparameters[j])
+        out[i] = updated_param1
+        out[j] = updated_param2
+        (index1, index2) = FindRepeatingValue(out[num_train_sets:])
+
+    # now parameters are validated and can be appended to the output set
+    for param in out[num_train_sets:]: 
+        out_param.append(format(param))
+
+    return ','.join(out_param)
+
+# this function checks repeating values in parameters and returns the index
+# of the first pair of repeating values if any is found 
+def FindRepeatingValue(parameters):
+    for i in range(0, len(parameters)):
+        for j in range(i + 1, len(parameters)):
+            if parameters[i] == parameters[j]:
+                return(i, j)
+    return (-1, -1)
+
+# this function rerounds repeating values until they are different and returns the values after rerounding.
+# if they are indeed the same (collison), it exits with a warning (this case can be very rare) 
+def AdjustValues(number1, number2, index1, index2, ref_number1, ref_number2):
+    x = number1
+    y = number2
+    # set a maximum number for iterations
+    round_max_digit = 20
+    # round to the next decimal place if found collison 
+    round_position = max(len(str(x)) - 2, len(str(y)) - 2) + 1
+    while x == y and round_position <= round_max_digit:
+        x = round(ref_number1, round_position)
+        y = round(ref_number2, round_position)
+        round_position += 1
+    # actually validate_metaparameters.py will validate those parameters later. 
+    # except for satisfying condition that those parameters should be in 
+    # range (0, 1), parameters D1, D2, D3, D4 for a certain xgram model 
+    # should satisfy D1 > D2 > D3 > D4. 
+    # here we only checked the repeating values but not the order relationship
+    if x == y:
+        sys.exit("train_lm.py: the {0}th and {1}th parameters are both {2}. metaparameters can not be exactly the same.".format(index1, index2, y))
```

Comment:
this could cause a crash for valid metaparameters if they are the same for different orders.  The relation D1>D2>D3>D4 is only within a particular order, there can be repeats for different orders.  You might need to reorganize the code a bit.